tox/config.py

```diff
@@ -791,9 +791,8 @@ def __init__(self, config, inipath):
             section = testenvprefix + name
             factors = set(name.split('-'))
             if section in self._cfg or factors <= known_factors:
-                config.envconfigs[name] = \
-                    self.make_envconfig(name, section, reader._subs, config,
-                                        replace=name in config.envlist)
```

Comment:
This list only contains the envs in envlist and would mean that only envconfigs get substituted that are defined in envlist. This is why we always replace at this early stage (configuration parsing) and don't crash if something is missing in a testenv. When an actual testrun is started we can check, if a testenv would be run, that has missing substitutions.